### ISSUE TITLE
BadImageFormatException on invocation of following code

### ISSUE THREAD

**Issue by JeroenBos** [dotnet/roslyn#16399]
**Version Used**:  15.0.26020.0
**Steps to Reproduce**:

1. Create a new console application.
2. Substitute the default code in the only .cs file for
```
using System;
using System.Collections.Generic;
using System.Linq;
public static class EnumerableExtensions
{
	static void Main(string[] args)
	{
		GetLeaves<object>(new List<object>(), list => null);
	}
	public static IEnumerable<T> GetLeaves<T>(T root, Func<T, IEnumerable<T>> getChildren)
	{
		return GetLeaves(root);

		IEnumerable<T> GetLeaves(T node)
		{
			var children = getChildren(node);
			if (children == null)
			{
				return null;
			}
			else
			{
				return children.SelectMany(GetLeaves);
			}
		}
	}
}

```

3. Run the program

I've reduced the code to this minimal example. Removal of any other code resulted in expected behavior.

**Expected Behavior**:

The program to terminate. 

**Actual Behavior**:

A `System.BadImageFormatException` is thrown at line `return GetLeaves(root);` Unfortunately, the exception does not provide much more details.

**Comment by nil4** [comment]
Confirmed using VS 2017. The exception is: 
```
System.BadImageFormatException occurred
  HResult=0x8007000B
  Message=An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
  StackTrace:
   at EnumerableExtensions.<>c__DisplayClass1_0`1.<GetLeaves>g__GetLeaves0(T node) in D:\projects\x\Program.cs:line 25
   at EnumerableExtensions.GetLeaves[T](T root, Func`2 getChildren) in D:\projects\x\Program.cs:line 12
   at EnumerableExtensions.Main(String[] args) in D:\projects\x\Program.cs:line 8
```

And running PEVerify on the executable shows:

```
Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

[IL]: Error: [D:\projects\x\bin\Debug\x.exe : EnumerableExtensions+<>c__DisplayClass1_0`1[T]::<GetLe
aves>g__GetLeaves0][offset 0x00000024] Unable to resolve token.
1 Error(s) Verifying D:\projects\x\bin\Debug\x.exe
```

**Comment by agocke** [comment]
No longer reproes.

I believe this was fixed by #16039

**Comment by agocke** [comment]
Closed by #16471
